giza/giza/operations/stage.py

```diff
@@ -350,34 +381,40 @@ def stage(self, root):
 
         return
 
-    def __upload(self, local_path, src_path, file_hash):
-        full_name = '/'.join((self.namespace, local_path))
-        k = boto.s3.key.Key(self.bucket)
 
-        try:
-            if local_path in self.SHARED_CACHE_BLACKLIST:
-                LOGGER.info('Uploading %s', local_path)
-                k.key = full_name
-                k.set_contents_from_filename(src_path, **self.S3_OPTIONS)
-                return None
+def _upload_file_to_s3(local_path, src_path, file_hash, namespace, bucket, should_retry=False):
+    full_name = '/'.join((namespace, local_path))
+    k = boto.s3.key.Key(bucket)
 
-            # Try to copy the file from the cache to its destination
-            file_hash = 'cache/' + binascii.b2a_hex(file_hash)
-            k.key = file_hash
+    try:
+        if local_path in Staging.SHARED_CACHE_BLACKLIST:
+            logger.info('Uploading %s', local_path)
+            k.key = full_name
+            k.set_contents_from_filename(src_path, **Staging.S3_OPTIONS)
+            return None
 
-            try:
-                k.copy(self.bucket.name, full_name, **self.S3_OPTIONS)
-            except boto.exception.S3ResponseError as err:
-                if err.status == 404:
-                    # Not found in cache. Upload it to the cache
-                    LOGGER.info('Uploading %s', local_path)
-                    k.set_contents_from_filename(src_path, **self.S3_OPTIONS)
-
-                    # And then copy it to its final destination
-                    k.copy(self.bucket.name, full_name, **self.S3_OPTIONS)
-                else:
-                    raise err
+        # Try to copy the file from the cache to its destination
+        file_hash = 'cache/' + binascii.b2a_hex(file_hash)
+        k.key = file_hash
+
+        try:
+            k.copy(bucket.name, full_name, **Staging.S3_OPTIONS)
         except boto.exception.S3ResponseError as err:
+            if err.status == 404:
+                # Not found in cache. Upload it to the cache
+                logger.info('Uploading %s', local_path)
+                k.set_contents_from_filename(src_path, **Staging.S3_OPTIONS)
+
+                # And then copy it to its final destination
+                k.copy(bucket.name, full_name, **Staging.S3_OPTIONS)
+            else:
+                raise err
+    except boto.exception.S3ResponseError as err:
+        if should_retry is True:
+            logger.error("encountered error uploading {0}, retrying once in a bit.".format())
```

Comment:
This should probably be log level "info" (or even "debug"), not "error". Getting this a couple times per stage is perfectly normal.